### PR TITLE
feat: osmosis provider

### PR DESCRIPTION
**Pull request by aljo242**
Closes: BLO-1537

```
{
  "prices": {
    "OSMO/USD": "54165419",
    "STARS/USD": "11621825831747928",
    "USDT/USD": "1000200040"
  },
  "timestamp": "2024-07-17T15:03:28.675080795Z"
}
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -283,6 +285,7 @@ gen-mocks:
 	@echo "--> generating mocks"
 	@go install github.com/vektra/mockery/v2
 	@go generate ./...
+	@cd ./providers/apis/defi/osmosis && go generate ./...
```
Comment (technicallyty) [inline Makefile]: shouldn't `go generate ./...` cover this?

File: pkg/http/round_tripper_with_headers.go [hunk 2]
```diff
@@ -58,3 +60,53 @@ func (r *RoundTripperWithHeaders) RoundTrip(req *http.Request) (*http.Response,
...
+		opt(req)
+	}
+
+	resp, err := c.internal.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("http get: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New(http.StatusText(resp.StatusCode))
```
Comment (technicallyty) [inline pkg/http/round_tripper_with_headers.go]: `resp.Status` contains `[code]: [status_text]`. could just use that here

File: providers/apis/defi/osmosis/client.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	SpotPrice(grpcCtx context.Context,
+		poolID uint64,
+		baseAsset,
+		quoteAsset string,
+	) (SpotPriceResponse, error)
+}
+
+// ClientImpl is an implementation of a client to Osmosis using a
+// poolmanager Query Client.
+type ClientImpl struct {
```
Comment (technicallyty) [inline providers/apis/defi/osmosis/client.go:36]: nit: seems like this and `MultiClientImpl` don't need to be exported types

File: providers/apis/defi/osmosis/client.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	}
+
+	wg.Wait()
+
+	return filterSpotPriceResponses(resps)
+}
+
+// filterSpotPriceResponses currently just chooses a random response as there is no way to differentiate.
+// TODO differentiate.
```
Comment (technicallyty) [inline providers/apis/defi/osmosis/client.go]: still todo?

File: providers/apis/defi/osmosis/price_fetcher.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+			if err != nil {
+				pf.logger.Error("failed parse spot price response", zap.Error(err))
+				unresolved[ticker] = providertypes.UnresolvedResult{
+					ErrorWithCode: providertypes.NewErrorWithCode(
+						err,
+						providertypes.ErrorFailedToParsePrice,
+					),
+				}
+				return
+			}
```
Comment (technicallyty) [inline providers/apis/defi/osmosis/price_fetcher.go]: shouldn't this have the unresolvedMtx lock?

maybe we should replace this logic with a closure:

```go
writeToUnresolved := func(ticker, err, code) {
    ....
}
```

File: providers/apis/defi/osmosis/client.go (hunk 4 above)
Comment (aljo242) [reply providers/apis/defi/osmosis/client.go]: fixed

File: providers/apis/defi/osmosis/price_fetcher.go (hunk 5 above)
Comment (technicallyty) [reply providers/apis/defi/osmosis/price_fetcher.go]: yo @aljo242 any thoughts on the closure idea above?

File: providers/apis/defi/osmosis/client.go (hunk 4 above)
Comment (technicallyty) [reply providers/apis/defi/osmosis/client.go]: comment still says TODO @aljo242

File: providers/apis/defi/osmosis/client.go (hunk 3 above)
Comment (technicallyty) [reply providers/apis/defi/osmosis/client.go:36]: this was marked as resolved, but there wasn't a change @aljo242 , i noticed the other files have their client impl as exported types as well. so maybe fine for now, but could be a ticket for the future

File: providers/apis/defi/osmosis/price_fetcher.go (hunk 5 above)
Comment (aljo242) [reply providers/apis/defi/osmosis/price_fetcher.go]: pushed here https://github.com/skip-mev/slinky/pull/607/commits/bb7b13dd2d1d9a118c645a18ae70d789dbc6e577

File: pkg/http/round_tripper_with_headers.go [hunk 6]
```diff
@@ -58,3 +60,53 @@ func (r *RoundTripperWithHeaders) RoundTrip(req *http.Request) (*http.Response,
...
+	if err != nil {
+		return nil, fmt.Errorf("http get: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New(resp.Status)
+	}
+
+	return resp, nil
+}
```
Comment (davidterpay) [inline pkg/http/round_tripper_with_headers.go:112]: Why do we need this additional client? We already have a separate client in base/api/handlers/requesthandler?

Comment (davidterpay) [reply pkg/http/round_tripper_with_headers.go:112]: can this functionality just be added there?

File: go.mod [hunk 7]
```diff
@@ -1,6 +1,6 @@
 module github.com/skip-mev/slinky
 
-go 1.22
+go 1.22.3
```
Comment (davidterpay) [inline go.mod]: can we make go verison bumps in separate PRs?

File: providers/apis/defi/osmosis/client.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+var (
+	_ Client = &ClientImpl{}
+	_ Client = &MultiClientImpl{}
+)
+
+// Client is the expected interface for an osmosis client.
+//
+//go:generate mockery --name Client --output ./mocks/ --case underscore
+type Client interface {
+	SpotPrice(grpcCtx context.Context,
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go]:
```suggestion
	SpotPrice(ctx context.Context,
```

File: providers/apis/defi/osmosis/types.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+	return tickerMetadata, nil
+}
+
+// DefaultAPIConfig is the default configuration for the Osmosis API price fetcher.
+var DefaultAPIConfig = config.APIConfig{
+	Enabled:          true,
+	Name:             Name,
+	Timeout:          4 * time.Second,
+	Interval:         1000 * time.Millisecond,
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/types.go]: interval can probably be higher than 1 sec since osmo block times are higher.

File: providers/apis/defi/osmosis/types.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	ReconnectTimeout: 4 * time.Second,
+	MaxQueries:       10,
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/types.go:122]: why do we need to allow 10 concurrent queries?

File: providers/apis/defi/osmosis/types.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+	Atomic:           false,
+	BatchSize:        25, // maximal # of accounts in getMultipleAccounts query is 100
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/types.go]: what is the batchsize doing?

File: providers/apis/defi/osmosis/client.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+	if err != nil {
+		return SpotPriceResponse{}, err
+	}
+
+	var spotPriceResponse SpotPriceResponse
+	if err := json.NewDecoder(resp.Body).Decode(&spotPriceResponse); err != nil {
+		return SpotPriceResponse{}, err
+	}
+
+	c.apiMetrics.AddRPCStatusCode(c.api.Name, c.redactedURL, metrics.RPCCodeOK)
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go]: why are we recording RPC status codes when an http client is used?

File: providers/apis/defi/osmosis/client.go [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+			// Observe the latency of the request.
+			start := time.Now()
+			defer func() {
+				wg.Done()
+				mc.apiMetrics.ObserveProviderResponseLatency(mc.api.Name, metrics.RedactedEndpointURL(index), time.Since(start))
+			}()
+
+			resp, err := client.SpotPrice(ctx, poolID, baseAsset, quoteAsset)
+			if err != nil {
+				mc.apiMetrics.AddRPCStatusCode(mc.api.Name, metrics.RedactedEndpointURL(index), metrics.RPCCodeError)
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go]: same as above

File: providers/apis/defi/osmosis/client.go [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+				mc.logger.Error("failed to spot price in sub client", zap.String("url", url), zap.Error(err))
+				return
+			}
+
+			mc.apiMetrics.AddRPCStatusCode(mc.api.Name, metrics.RedactedEndpointURL(index), metrics.RPCCodeOK)
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go]: same as above

File: providers/apis/defi/osmosis/client.go [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+	perm := rand.Perm(len(responses))
+	for _, i := range perm {
+		resp := responses[perm[i]]
+		if resp.SpotPrice != "" {
+			return resp, nil
+		}
+	}
+
+	return SpotPriceResponse{}, fmt.Errorf("no responses found")
+}
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go:230]: if we are using multi-client doesn't it make more sense to also use query block height?

Comment (davidterpay) [reply providers/apis/defi/osmosis/client.go:230]: The eth and sol providers both do this.

File: providers/apis/defi/osmosis/price_fetcher.go [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+	ctx context.Context,
+	tickers []oracletypes.ProviderTicker,
+) oracletypes.PriceResponse {
+	resolved := make(oracletypes.ResolvedPrices)
+	unresolved := make(oracletypes.UnResolvedPrices)
+
+	wg := sync.WaitGroup{}
+	unresolvedMtx := sync.Mutex{}
+	resolveMtx := sync.Mutex{}
+	wg.Add(len(tickers))
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/price_fetcher.go]: This will spin up an unbounded number of threads to make concurrent requests?

File: providers/apis/defi/osmosis/price_fetcher.go [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+				))
+
+				return
+			}
+
+			resolvedTickerCallback(ticker, price)
+		}(ticker)
+	}
+
+	wg.Wait()
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/price_fetcher.go]: this should ideally be buffering a maximum number of concurrent requests otherwise we will certainly be rate limited.

File: pkg/http/round_tripper_with_headers.go (hunk 6 above)
Comment (aljo242) [reply pkg/http/round_tripper_with_headers.go:112]: this is an internal general http client and doesnt need all of the requestHandler functionality.  It makes sense to me to only have the Fetcher use a client that gives it the functionality that it needs

File: providers/apis/defi/osmosis/client.go (hunk 15 above)
Comment (aljo242) [reply providers/apis/defi/osmosis/client.go:230]: This query does not return a block height so we would have to do an additional query alongside what we're doing

File: providers/apis/defi/osmosis/price_fetcher.go (hunk 17 above)
Comment (aljo242) [reply providers/apis/defi/osmosis/price_fetcher.go]: Now using an errgroup that limits the number of goroutines to `MaxQueries`

**Review by technicallyty — APPROVED** [review]
lgtm

File: providers/apis/defi/osmosis/client.go [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+	for i, client := range mc.clients {
+		url := mc.api.Endpoints[i].URL
+
+		index := i
+		go func(index int, client Client) {
+			// Observe the latency of the request.
+			start := time.Now()
+			defer func() {
+				wg.Done()
+				mc.apiMetrics.ObserveProviderResponseLatency(mc.api.Name, metrics.RedactedEndpointURL(index), time.Since(start))
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go]: is this metric going to be duplicated since the individual clients themselves already record this?

File: providers/apis/defi/osmosis/client.go [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+		}(index, client)
+
+	}
+
+	wg.Wait()
+
+	return filterSpotPriceResponses(resps)
+}
+
+// filterSpotPriceResponses currently just chooses a random response as there is no way to differentiate.
```
Comment (davidterpay) [inline providers/apis/defi/osmosis/client.go:215]: should probably make a todo task out of this to see whether we want to also query block heights.

**Review by davidterpay — APPROVED** [review]
small nit
